### PR TITLE
Simplify Replica Allocator

### PR DESCRIPTION
**Pull request by kimchy**
Simplify the codebase of replica allocator and add more unit tests for it

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java [hunk 1]
```diff
@@ -203,5 +125,134 @@ public boolean allocateUnassigned(RoutingAllocation allocation) {
...
+    /**
+     * Finds the store for the assigned shard in the fetched data, returns null if none is found.
+     */
+    private TransportNodesListShardStoreMetaData.StoreFilesMetaData findStore(ShardRouting shard, RoutingAllocation allocation, AsyncShardFetch.FetchResult<TransportNodesListShardStoreMetaData.NodeStoreFilesMetaData> data) {
+        assert shard.currentNodeId() != null;
+        DiscoveryNode primaryNode = allocation.nodes().get(shard.currentNodeId());
+        if (primaryNode == null) {
+            return null;
+        }
+        TransportNodesListShardStoreMetaData.NodeStoreFilesMetaData primaryNodeFilesStore = data.getData().get(primaryNode);
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java]: Out of curiosity, does this need to check `data.hasData()` first?

Comment (dakrone) [inline core/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java]: Ah nevermind, I see where we check it above :)

File: core/src/test/java/org/elasticsearch/gateway/ReplicaShardAllocatorTests.java [hunk 2]
```diff
@@ -194,15 +196,42 @@ public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, Routing
...
+            // we sometime return empty list of files, make sure we test this as well
+            testAllocator.addData(node2, false, null);
+        }
+        boolean changed = testAllocator.allocateUnassigned(allocation);
+        assertThat(changed, equalTo(true));
+        assertThat(allocation.routingNodes().unassigned().ignored().size(), equalTo(1));
+        assertThat(allocation.routingNodes().unassigned().ignored().get(0).shardId(), equalTo(shardId));
+
+        allocation = onePrimaryOnNode1And1Replica(yesAllocationDeciders(),
+                Settings.builder().put(UnassignedInfo.INDEX_DELAYED_NODE_LEFT_TIMEOUT_SETTING, TimeValue.timeValueHours(1)).build(), UnassignedInfo.Reason.NODE_LEFT);
```
Comment (dakrone) [inline core/src/test/java/org/elasticsearch/gateway/ReplicaShardAllocatorTests.java]: Maybe I don't fully understand the delayed allocation, does it only happen for primaries?

**Comment by dakrone** [comment]
LGTM, I like the encapsulation, it's much easier to read! Left one question (not blocking).

File: core/src/test/java/org/elasticsearch/gateway/ReplicaShardAllocatorTests.java (hunk 2 above)
Comment (kimchy) [inline core/src/test/java/org/elasticsearch/gateway/ReplicaShardAllocatorTests.java]: the way we implemented it is that it only applies to replica shards, in practice, that is what we need to delay really

**Pull request merged**